File: kronos/core/Mon.py
from pymongo import MongoClient
from pymongo.errors import DuplicateKeyError
from typing import List, Dict, Union

class Mon:
    def __init__(self, db_name: str, host: str = "localhost", port: int = 27017):
        self.client = MongoClient(host, port)
        self.db = self.client[db_name]

    def inse(self, collection: str, document: dict) -> Union[int, bool, None]:
        """Insert a document into a collection."""
        try:
            result = self.db[collection].insert_one(document)
            return result.inserted_id if result.inserted_id else True
        except DuplicateKeyError:
            print("Duplicate entry found. Entry was not added.")
            return False
        except Exception as e:
            print(f"Database error occurred: {e}")
            return None

    def q(self, collection: str, query: dict, update: dict = None) -> bool:
        """Insert, update or delete (handled by MongoDB's replace_one/update_one/delete_one)."""
        if not update:
            print("Missing update details for the query.")
            return False

        try:
            result = self.db[collection].update_one(query, {"$set": update})
            return result.modified_count > 0
        except Exception as e:
            print(f"Error executing update query: {e}")
            return False

    def f(self, collection: str, query: dict) -> Union[Dict, str, bool]:
        """Fetch a single document based on the query."""
        try:
            result = self.db[collection].find_one(query)
            return result if result else False
        except Exception as e:
            print(f"Error fetching document: {e}")
            return False

    def fa(self, collection: str, query: dict) -> Union[List[Dict], bool]:
        """Fetch multiple documents."""
        try:
            result = list(self.db[collection].find(query))
            return result if result else False
        except Exception as e:
            print(f"Error fetching documents: {e}")
            return False

    def fl(self, fields: Union[str, List[str]], collection: str, query: dict = {}) -> Union[Dict, bool]:
        """Fetch specific fields and return either a row or key-value pairs."""
        try:
            projection = {field: 1 for field in fields} if isinstance(fields, list) else {fields: 1}
            result = self.db[collection].find(query, projection)

            if isinstance(fields, list) and len(fields) == 2:
                # Couple List (key-value pairs)
                return {doc[fields[0]]: doc[fields[1]] for doc in result if fields[0] in doc and fields[1] in doc}
            else:
                # Row List
                return [doc[fields] for doc in result if fields in doc]
        except Exception as e:
            print(f"Error fetching specific fields: {e}")
            return False

File: gaia/core/Mon.php
<?php
namespace Core;
use MongoDB\Client;
use MongoDB\BSON\ObjectId;
use MongoDB\Driver\Exception\Exception;
/**
TODO Solve undefined MongoDB\Client
*/
class Mon {
    protected $uri="mongodb://dros:130177@127.0.0.1:27017/?authMechanism=SCRAM-SHA-1&authSource=admin";
    protected $client;
    protected $db;

    public function __construct(string $dbName) {
       try {
           $this->client = new Client($this->uri);
       } catch (Exception $e) {
           echo "Failed to connect to MongoDB: " . $e->getMessage();
       }
        $this->db = $this->client->selectDatabase($dbName);
    }
    // List all collections in the database
    protected function listCollections(): array {
        try {
            $collections = $this->db->listCollections();
            return array_map(fn($collection) => $collection->getName(), iterator_to_array($collections));
        } catch (Exception $e) {
            echo "MongoDB error occurred while listing collections: " . $e->getMessage();
            return [];
        }
    }
  // Fetch multiple documents
    protected function fa(string $collectionName, array $filter = []): array|bool
    {
        try {
            $collection = $this->db->selectCollection($collectionName);
            $cursor = $collection->find($filter);

            $documents = iterator_to_array($cursor);
            return !empty($documents) ? $documents : false;
        } catch (Exception $e) {
            echo "MongoDB error occurred: " . $e->getMessage();
            return false;
        }
    }

    // Fetch rows (one column) or key-value pairs (two columns)
    protected function fl(string|array $fields, string $collectionName, array $filter = []): array|bool
    {
        try {
            $projection = [];

            // Single field projection
            if (is_string($fields)) {
                $projection = [$fields => 1];
            }
            // Couple list projection (key-value pairs)
            elseif (is_array($fields) && count($fields) == 2) {
                $projection = [$fields[0] => 1, $fields[1] => 1];
            }

            $collection = $this->db->selectCollection($collectionName);
            $cursor = $collection->find($filter, ['projection' => $projection]);

            $result = [];
            if (is_string($fields)) {
                foreach ($cursor as $doc) {
                    $result[] = $doc[$fields];
                }
            } elseif (count($fields) == 2) {
                foreach ($cursor as $doc) {
                    $result[$doc[$fields[0]]] = $doc[$fields[1]];
                }
            }

            return !empty($result) ? $result : false;
        } catch (Exception $e) {
            echo "MongoDB error occurred: " . $e->getMessage();
            return false;
        }
    }
    // Insert a document into a collection
    protected function inse(string $collectionName, array $document): ObjectId|bool|null
    {
        try {
            $collection = $this->db->selectCollection($collectionName);
            $result = $collection->insertOne($document);

            // Return the inserted ID if successful
            return $result->getInsertedId() ?: true;
        } catch (Exception $e) {
            if ($e->getCode() == 11000) {
                echo "Duplicate entry found. Entry was not added.";
            } else {
                echo "MongoDB error occurred: " . $e->getMessage();
            }
            return false;
        }
    }

    // Update a document in a collection
    protected function q(string $collectionName, array $filter, array $update): bool
    {
        try {
            $collection = $this->db->selectCollection($collectionName);
            $result = $collection->updateOne($filter, ['$set' => $update]);

            return $result->getModifiedCount() > 0;
        } catch (Exception $e) {
            echo "MongoDB error occurred: " . $e->getMessage();
            return false;
        }
    }

    // Fetch a single document based on a query
    protected function f(string $collectionName, array $filter): array|bool
    {
        try {
            $collection = $this->db->selectCollection($collectionName);
            $result = $collection->findOne($filter);

            return $result ? $result->getArrayCopy() : false;
        } catch (Exception $e) {
            echo "MongoDB error occurred: " . $e->getMessage();
            return false;
        }
    }

}
